tls/s2n_prf.c

```diff
@@ -78,38 +79,131 @@ static int s2n_sslv3_prf(union s2n_prf_working_space *ws, struct s2n_blob *secre
     return 0;
 }
 
-static int s2n_p_hash(union s2n_prf_working_space *ws, s2n_hmac_algorithm alg, struct s2n_blob *secret,
+static int s2n_p_hash_init(union s2n_prf_working_space *ws, s2n_hmac_algorithm alg, const struct s2n_blob *secret)
+{
+    int r = 0;
+
+    eq_check(ws->tls.mac_key, NULL);
+    notnull_check(ws->tls.mac_key = EVP_PKEY_new_mac_key(EVP_PKEY_HMAC, NULL, secret->data, secret->size));
+
+    eq_check(ws->tls.md_ctx, NULL);
+#if S2N_OPENSSL_VERSION_AT_LEAST(1,1,0) && !defined(LIBRESSL_VERSION_NUMBER)
+    notnull_check(ws->tls.md_ctx = EVP_MD_CTX_new());
+#else
+    notnull_check(ws->tls.md_ctx = EVP_MD_CTX_create());
+#endif
+
+    EVP_MD_CTX_set_flags(ws->tls.md_ctx, EVP_MD_CTX_FLAG_NON_FIPS_ALLOW);
+
+    switch (alg) {
+    case S2N_HMAC_SSLv3_MD5:
+    case S2N_HMAC_MD5:
+        r = EVP_DigestSignInit(ws->tls.md_ctx, NULL, EVP_md5(), NULL, ws->tls.mac_key);
+        break;
+    case S2N_HMAC_SSLv3_SHA1:
+    case S2N_HMAC_SHA1:
+        r = EVP_DigestSignInit(ws->tls.md_ctx, NULL, EVP_sha1(), NULL, ws->tls.mac_key);
+        break;
+    case S2N_HMAC_SHA224:
+        r = EVP_DigestSignInit(ws->tls.md_ctx, NULL, EVP_sha224(), NULL, ws->tls.mac_key);
+        break;
+    case S2N_HMAC_SHA256:
+        r = EVP_DigestSignInit(ws->tls.md_ctx, NULL, EVP_sha256(), NULL, ws->tls.mac_key);
+        break;
+    case S2N_HMAC_SHA384:
+        r = EVP_DigestSignInit(ws->tls.md_ctx, NULL, EVP_sha384(), NULL, ws->tls.mac_key);
+        break;
+    case S2N_HMAC_SHA512:
+        r = EVP_DigestSignInit(ws->tls.md_ctx, NULL, EVP_sha512(), NULL, ws->tls.mac_key);
+        break;
+    default:
+        S2N_ERROR(S2N_ERR_P_HASH_INVALID_ALGORITHM);
+    }
+
+    if (r == 0) {
+        S2N_ERROR(S2N_ERR_P_HASH_INIT_FAILED);
+    }
+
+    return 0;
+}
+
+static int s2n_p_hash_update(EVP_MD_CTX *md_ctx, const void *data, size_t size)
+{
+    if (EVP_DigestSignUpdate(md_ctx, data, size) == 0) {
+        S2N_ERROR(S2N_ERR_P_HASH_UPDATE_FAILED);
+    }
+
+    return 0;
+}
+
+static int s2n_p_hash_final(EVP_MD_CTX *md_ctx, unsigned char *digest, size_t *size)
+{
+    if (EVP_DigestSignFinal(md_ctx, digest, size) == 0) {
+        S2N_ERROR(S2N_ERR_P_HASH_FINAL_FAILED);
+    }
+
+    return 0;
+}
+
+static int s2n_p_hash_free(union s2n_prf_working_space *ws)
+{
+    notnull_check(ws->tls.mac_key);
+    notnull_check(ws->tls.md_ctx);
+
+    EVP_PKEY_free(ws->tls.mac_key);
```

Comment:
The comment above about the equivalence of `EVP_MD_CTX_new` and `EVP_MD_CTX_create` also applies for `EVP_MD_CTX_free` and `EVP_MD_CTX_destroy`